src/main/java/org/opensearch/neuralsearch/processor/optimization/InferenceFilter.java

```diff
@@ -0,0 +1,233 @@
+     * @param embeddingList The list of existing embeddings.
+     * @param sourceAndMetadataMap The metadata map of the new document.
+     * @param existingSourceAndMetadataMap The metadata map of the existing document.
+     * @param fullEmbeddingKey The dot-notation path for the embedding field.
+     * @return A processed list or an empty list if embeddings are reused.
+     */
+
+    public abstract List<Object> filterInferenceValuesInList(
+        List<Object> processList,
+        List<Object> existingList,
+        List<Object> embeddingList,
+        Map<String, Object> sourceAndMetadataMap,
+        Map<String, Object> existingSourceAndMetadataMap,
+        String fullEmbeddingKey
+    );
+
+    /**
+     * This method navigates through the nested structure, checking each key-value pair recursively. It supports:
+     * Map values: Processed recursively using this method.
+     * List values: Processed using filterListValue.
+     * Primitive values: Directly compared using filterInferenceValue.
+     *
+     * @param existingSourceAndMetadataMap The metadata map of the existing document.
+     * @param sourceAndMetadataMap The metadata map of the new document.
+     * @param processMap The current map being processed.
+     *
+     * @return A filtered map containing only elements that require new embeddings.
+     *
+     */
+    public Map<String, Object> filter(
+        Map<String, Object> existingSourceAndMetadataMap,
+        Map<String, Object> sourceAndMetadataMap,
+        Map<String, Object> processMap
+    ) {
+        return filter(existingSourceAndMetadataMap, sourceAndMetadataMap, processMap, "");
+    }
+
+    /**
+     * Helper method for filter
+     * @param existingSourceAndMetadataMap The metadata map of the existing document.
+     * @param sourceAndMetadataMap The metadata map of the new document.
+     * @param processMap The current map being processed.
+     * @param traversedPath The dot-notation path of the previously traversed elements.
+     * e.g:
+     * In a map structured like:
+     *       level1
+     *         level2
+     *            level3
+     * traversedPath would be dot-separated string of level1.level2.level3
+     * @return A filtered map containing only elements that require new embeddings
+     */
+    private Map<String, Object> filter(
+        Map<String, Object> existingSourceAndMetadataMap,
+        Map<String, Object> sourceAndMetadataMap,
+        Object processMap,
+        String traversedPath
+    ) {
+        if (processMap instanceof Map == false) {
+            throw new IllegalArgumentException("processMap needs to be an instanceof Map");
+        }
+        Map<String, Object> filteredProcessMap = new HashMap<>();
+        Map<String, Object> castedProcessMap = ProcessorUtils.castToMap(processMap);
+        for (Map.Entry<?, ?> entry : castedProcessMap.entrySet()) {
+            if ((entry.getKey() instanceof String) == false) {
```

Comment:
You don't need this check. castedProcessMap is already Map<String, Object> and the key is String.